proposals/Simple-programs.md

```diff
@@ -0,0 +1,169 @@
+# Simple programs
+
+* [x] Proposed
+* [x] Prototype: Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Allow a sequence of *statements* to occur right before the *namespace_member_declaration*s of a *compilation_unit* (i.e. source file).
+
+The semantics are that if such a sequence of *statements* is present, the following type declaration, modulo the actual type name and the method name, would be emitted:
+
+``` c#
+static class Program
+{
+    static async Task Main()
+    {
+        // statements
+    }
+}
+```
+
+See also https://github.com/dotnet/csharplang/issues/3117.
+
+## Motivation
+[motivation]: #motivation
+
+There's a certain amount of boilerplate surrounding even the simplest of programs,
+because of the need for an explicit `Main` method. This seems to get in the way of
+language learning and program clarity. The primary goal of the feature therefore is
+to allow C# programs without unnecessary boilerplate around them, for the sake of
+learners and the clarity of code.
+
+## Detailed design
+[design]: #detailed-design
+
+### Syntax
+
+The only additional syntax is allowing a sequence of *statement*s in a compilation unit,
+just before the *namespace_member_declaration*s:
+
+``` antlr
+compilation_unit
+    : extern_alias_directive* using_directive* global_attributes? statement* namespace_member_declaration*
+    ;
+```
+
+In all but one *compilation_unit* the *statement*s must all be local function declarations. 
```

Comment:
Should there even be multiple compilation units with statements to support local functions? Or can this be covered by scenario 2 (top level functions) later?

Reply:
> Should there even be multiple compilation units with statements to support local functions?

There is no requirement to place Local Functions into a separate compilation unit.

---
In reply to: [375677410](https://github.com/dotnet/csharplang/pull/3159#discussion_r375677410) [](ancestors = 375677410)